src/cart/crt_rpc.c

```diff
@@ -854,24 +854,27 @@ uri_lookup_cb(const struct crt_cb_info *cb_info)
 		D_GOTO(out, rc);
 	}
 
-	/* After a URI lookup, check if membership list has this rank.
-	 * If not - we discovered a new rank and need to populate it in membs
-	 * list of the group.
+	/* After a URI lookup, if we are not a service (whose groups are
+	 * strictly versioned), check if membership list has this rank. If not
+	 * - we discovered a new rank and need to populate it in membs list of
+	 * the group.
 	 */
-	D_RWLOCK_WRLOCK(&grp_priv->gp_rwlock);
-	membs = grp_priv_get_membs(grp_priv);
-	found = d_rank_list_find(membs, ul_in->ul_rank, NULL);
+	if (!crt_is_service()) {
```

Comment:
this looks fine. a client might end up not knowing about a rank and sending an rpc to it, resulting in a uri_lookup to psr, but a server at most would only do a uri_lookups on known ranks, but for tags that it does not know about (and only for non-contig providers such as psm2).
So in practice server should never discover a new rank via this logic:)